internal/infrastructure/kubernetes/infra_resource.go

```diff
@@ -74,8 +78,52 @@ func (i *Infra) createOrUpdateDeployment(ctx context.Context, r ResourceRender)
 		Name:      deployment.Name,
 	}
 
+	hpa, err := r.HorizontalPodAutoscaler()
+	if err != nil {
+		return err
+	}
+
+	var opts cmp.Options
+	if hpa != nil {
+		opts = append(opts, cmpopts.IgnoreFields(appsv1.DeploymentSpec{}, "Replicas"))
+	}
+
 	return i.Client.CreateOrUpdate(ctx, key, current, deployment, func() bool {
-		return !reflect.DeepEqual(deployment.Spec, current.Spec)
+		return !cmp.Equal(current.Spec, deployment.Spec, opts...)
+	})
+}
+
+// configureHPA configures HorizontalPodAutoscaler object in the kube api server based on the provided ResourceRender,
+// there are two operations on this method:
+// 1. if an HPA is defined and it doesn't exist then creates, or otherwise updates
+// 2. if an HPA is not defined then delete the object irrespective of its existence
+func (i *Infra) configureHPA(ctx context.Context, r ResourceRender) error {
```

Comment:
let's still call it `createOrUpdateHPA` for consistency

Reply:
but the operation also include deletion when needed, like omitting `envoyHpa`, wdyt? @shawnh2

Reply:
sure, lets keep it

Reply:
Why not move it into `deleteHPA`  lline 203 - 205 ?

Reply:
Oh, hold on, I just noticed, what if I implement this in the deleteHPA function? Do you think it would be suitable, @Xunzhuo ?

```go
func (i *Infra) deleteHPA(ctx context.Context, r ResourceRender) error {
	hpa, err := r.HorizontalPodAutoscaler()
	if err != nil {
		return err
	}

	if hpa == nil {
		hpa = &autoscalingv2.HorizontalPodAutoscaler{
			ObjectMeta: metav1.ObjectMeta{
				Namespace: i.Namespace,
				Name:      r.Name(),
			},
		}
	}

	return i.Client.Delete(ctx, hpa)
}
```

Reply:
Yes, I think this is more suitable, and configuteHPA can be focused on createOrUpdateHPA

Reply:
make sense, let me update it @Xunzhuo